File: backend/models/monitor.py
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from backend.db.database import Base


class AgentRun(Base):
    __tablename__ = "agent_runs"

    run_id: Mapped[str] = mapped_column(
        String(64),
        primary_key=True,
    )

    thread_id: Mapped[str] = mapped_column(
        String(128),
        index=True,
        nullable=False,
    )

    user_id: Mapped[int] = mapped_column(
        Integer,
        index=True,
        nullable=False,
    )

    message_id: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
    )

    input: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )

    output: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    status: Mapped[str] = mapped_column(
        String(32),
        default="running",
        nullable=False,
        index=True,
    )

    main_skill: Mapped[str | None] = mapped_column(
        String(128),
        nullable=True,
    )

    started_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.now,
        nullable=False,
        index=True,
    )

    finished_at: Mapped[datetime | None] = mapped_column(
        DateTime,
        nullable=True,
    )


class AgentRunStep(Base):
    __tablename__ = "agent_run_steps"

    id: Mapped[int] = mapped_column(
        Integer,
        primary_key=True,
        autoincrement=True,
    )

    run_id: Mapped[str] = mapped_column(
        String(64),
        ForeignKey("agent_runs.run_id", ondelete="CASCADE"),
        index=True,
        nullable=False,
    )

    step_index: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )

    node_name: Mapped[str | None] = mapped_column(
        String(128),
        index=True,
        nullable=True,
    )

    event_type: Mapped[str] = mapped_column(
        String(128),
        index=True,
        nullable=False,
    )

    input_json: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    output_json: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    error: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.now,
        nullable=False,
        index=True,
    )


class AgentToolCall(Base):
    __tablename__ = "agent_tool_calls"

    id: Mapped[int] = mapped_column(
        Integer,
        primary_key=True,
        autoincrement=True,
    )

    run_id: Mapped[str] = mapped_column(
        String(64),
        ForeignKey("agent_runs.run_id", ondelete="CASCADE"),
        index=True,
        nullable=False,
    )

    step_index: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )

    tool_name: Mapped[str] = mapped_column(
        String(255),
        index=True,
        nullable=False,
    )

    skill: Mapped[str | None] = mapped_column(
        String(128),
        index=True,
        nullable=True,
    )

    tool_input_json: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    tool_output_json: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    status: Mapped[str] = mapped_column(
        String(32),
        default="running",
        nullable=False,
        index=True,
    )

    latency_ms: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.now,
        nullable=False,
        index=True,
    )

File: agent-frontend/src/types/monitor.ts
export interface MonitorSummary {
  total_runs: number
  success_runs: number
  failed_runs: number
  running_runs: number
  success_rate: number
  avg_latency_ms: number
}

export interface AgentRun {
  run_id: string
  thread_id: string
  user_id: number
  username?: string | null
  user_label?: string | null
  message_id?: number | null
  input: string
  output?: string | null
  status: string
  main_skill?: string | null
  started_at: string
  finished_at?: string | null
  latency_ms?: number | null
}

export interface AgentRunStep {
  id: number
  run_id: string
  step_index: number
  node_name?: string | null
  event_type: string
  input_json?: string | null
  output_json?: string | null
  error?: string | null
  created_at: string
}

export interface AgentToolCall {
  id: number
  run_id: string
  step_index: number
  tool_name: string
  skill?: string | null
  tool_input_json?: string | null
  tool_output_json?: string | null
  status: string
  latency_ms?: number | null
  created_at: string
}

export interface AgentRunDetail {
  run: AgentRun
  steps: AgentRunStep[]
  tool_calls: AgentToolCall[]
  trace?: MonitorTraceItem[]
}

export interface MonitorToolTrace {
  id: number
  step_index: number
  tool_name: string
  skill?: string | null
  status: string
  latency_ms?: number | null
  input?: unknown
  output?: unknown
}

export interface MonitorTraceItem {
  id: string
  type: 'user_input' | 'router' | 'worker' | 'tool' | 'summarizer' | string
  title: string
  status?: string | null
  skill?: string | null
  step_index?: number | null
  end_step_index?: number | null
  input?: unknown
  output?: unknown
  tools?: MonitorToolTrace[]
  error?: string | null
}
